File: packages/python/src/codespar/agent_gate.py
"""
AgentGate type-narrowed helpers — Python parallel of agent-gate.ts.

Mirrors the TypeScript surface 1:1: five frozen dataclasses, five
discriminant constants, the ``AgentGateCode`` Literal union, the
``AGENT_GATE_CODES`` frozenset, five PEP 647 ``TypeGuard``
predicates, and ``assert_exhaustive_agent_gate``.

Each guard checks both the ``code`` discriminant against
``AGENT_GATE_CODES`` AND its own required sibling fields. A
payload with a well-formed ``code`` but a missing sibling returns
False rather than narrowing positive on the discriminant alone —
the test suite enforces this invariant per guard.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Final, Literal, TypeGuard

# ── discriminant constants ────────────────────────────────────────

POLICY_DENIED: Final[Literal["policy_denied"]] = "policy_denied"
APPROVAL_REQUIRED: Final[Literal["approval_required"]] = "approval_required"
MOCKS_EXHAUSTED: Final[Literal["mocks_exhausted"]] = "mocks_exhausted"
MOCKS_ENGINE_ERROR: Final[Literal["mocks_engine_error"]] = "mocks_engine_error"
TOOL_NOT_MOCKED: Final[Literal["tool_not_mocked"]] = "tool_not_mocked"

AgentGateCode = Literal[
    "policy_denied",
    "approval_required",
    "mocks_exhausted",
    "mocks_engine_error",
    "tool_not_mocked",
]

AGENT_GATE_CODES: Final[frozenset[str]] = frozenset(
    [
        POLICY_DENIED,
        APPROVAL_REQUIRED,
        MOCKS_EXHAUSTED,
        MOCKS_ENGINE_ERROR,
        TOOL_NOT_MOCKED,
    ]
)


# ── output dataclasses (frozen, slotted) ──────────────────────────


@dataclass(slots=True, frozen=True)
class PolicyDeniedOutput:
    rule_id: str
    message: str
    code: Literal["policy_denied"] = POLICY_DENIED


@dataclass(slots=True, frozen=True)
class ApprovalRequiredOutput:
    approval_id: str
    expires_at: str
    message: str
    code: Literal["approval_required"] = APPROVAL_REQUIRED


@dataclass(slots=True, frozen=True)
class MocksExhaustedOutput:
    message: str
    code: Literal["mocks_exhausted"] = MOCKS_EXHAUSTED


@dataclass(slots=True, frozen=True)
class MocksEngineErrorOutput:
    message: str
    code: Literal["mocks_engine_error"] = MOCKS_ENGINE_ERROR


@dataclass(slots=True, frozen=True)
class ToolNotMockedOutput:
    tool_name: str
    message: str
    code: Literal["tool_not_mocked"] = TOOL_NOT_MOCKED


AgentGateToolResultOutput = (
    PolicyDeniedOutput
    | ApprovalRequiredOutput
    | MocksExhaustedOutput
    | MocksEngineErrorOutput
    | ToolNotMockedOutput
)


# ── guards ────────────────────────────────────────────────────────


def _is_object(value: Any) -> TypeGuard[dict[str, Any]]:
    return isinstance(value, dict)


def _has_str(obj: dict[str, Any], key: str) -> bool:
    return isinstance(obj.get(key), str)


def is_policy_denied(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != POLICY_DENIED or code not in AGENT_GATE_CODES:
        return False
    return _has_str(value, "rule_id") and _has_str(value, "message")


def is_approval_required(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != APPROVAL_REQUIRED or code not in AGENT_GATE_CODES:
        return False
    return (
        _has_str(value, "approval_id")
        and _has_str(value, "expires_at")
        and _has_str(value, "message")
    )


def is_mocks_exhausted(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != MOCKS_EXHAUSTED or code not in AGENT_GATE_CODES:
        return False
    return _has_str(value, "message")


def is_mocks_engine_error(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != MOCKS_ENGINE_ERROR or code not in AGENT_GATE_CODES:
        return False
    return _has_str(value, "message")


def is_tool_not_mocked(value: Any) -> TypeGuard[dict[str, Any]]:
    if not _is_object(value):
        return False
    code = value.get("code")
    if code != TOOL_NOT_MOCKED or code not in AGENT_GATE_CODES:
        return False
    return _has_str(value, "tool_name") and _has_str(value, "message")


def assert_exhaustive_agent_gate(value: Any) -> None:
    """Raise on any AgentGate payload not matched by the five guards.

    Equivalent to TypeScript's ``assertExhaustiveAgentGate(value: never)``
    — call from a default branch after handling each variant so a sixth
    code landing without a handler trips at the boundary instead of
    being silently swallowed.
    """
    raise AssertionError(
        f"agent-gate: unexpected output variant {value!r}",
    )

File: packages/core/src/agent-gate.ts
/**
 * AgentGate type-narrowed helpers.
 *
 * Pure consumer-side narrowing for the hosted-test-mode tool_result
 * discriminated union (Backend D3). The five variants surface as
 * `tool_result.output` payloads on streamed `ToolCallRecord` values;
 * the guards turn `unknown` into one of the five `*Output`
 * interfaces so callers can branch without casting.
 *
 * Each guard checks both the `code` discriminant against
 * AGENT_GATE_CODES AND its own required sibling fields — so a
 * well-formed `code` with a missing sibling returns false rather
 * than narrowing positive on the discriminant alone. The exhaustive-
 * match utility (`assertExhaustiveAgentGate`) makes a switch over
 * AgentGateCode fail to compile if a sixth variant lands without
 * the consumer updating their handler.
 */

import type { ToolCallRecord } from "@codespar/types";

export const AgentGateCode = {
  PolicyDenied: "policy_denied",
  ApprovalRequired: "approval_required",
  MocksExhausted: "mocks_exhausted",
  MocksEngineError: "mocks_engine_error",
  ToolNotMocked: "tool_not_mocked",
} as const;

export type AgentGateCode = (typeof AgentGateCode)[keyof typeof AgentGateCode];

export const AGENT_GATE_CODES: ReadonlySet<AgentGateCode> = new Set([
  AgentGateCode.PolicyDenied,
  AgentGateCode.ApprovalRequired,
  AgentGateCode.MocksExhausted,
  AgentGateCode.MocksEngineError,
  AgentGateCode.ToolNotMocked,
]);

export interface PolicyDeniedOutput {
  code: typeof AgentGateCode.PolicyDenied;
  rule_id: string;
  message: string;
}

export interface ApprovalRequiredOutput {
  code: typeof AgentGateCode.ApprovalRequired;
  approval_id: string;
  expires_at: string;
  message: string;
}

export interface MocksExhaustedOutput {
  code: typeof AgentGateCode.MocksExhausted;
  message: string;
}

export interface MocksEngineErrorOutput {
  code: typeof AgentGateCode.MocksEngineError;
  message: string;
}

export interface ToolNotMockedOutput {
  code: typeof AgentGateCode.ToolNotMocked;
  tool_name: string;
  message: string;
}

export type AgentGateToolResultOutput =
  | PolicyDeniedOutput
  | ApprovalRequiredOutput
  | MocksExhaustedOutput
  | MocksEngineErrorOutput
  | ToolNotMockedOutput;

// Narrowed ToolCallRecord aliases — when a guard succeeds the
// `output` field is known to be the corresponding *Output variant.
export type PolicyDeniedToolCall = ToolCallRecord & {
  output: PolicyDeniedOutput;
};
export type ApprovalRequiredToolCall = ToolCallRecord & {
  output: ApprovalRequiredOutput;
};
export type MocksExhaustedToolCall = ToolCallRecord & {
  output: MocksExhaustedOutput;
};
export type MocksEngineErrorToolCall = ToolCallRecord & {
  output: MocksEngineErrorOutput;
};
export type ToolNotMockedToolCall = ToolCallRecord & {
  output: ToolNotMockedOutput;
};

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function readStringField(obj: Record<string, unknown>, key: string): string | null {
  const v = obj[key];
  return typeof v === "string" ? v : null;
}

export function isPolicyDenied(value: unknown): value is PolicyDeniedOutput {
  if (!isObject(value)) return false;
  if (value.code !== AgentGateCode.PolicyDenied) return false;
  if (!AGENT_GATE_CODES.has(value.code as AgentGateCode)) return false;
  return readStringField(value, "rule_id") !== null
    && readStringField(value, "message") !== null;
}

export function isApprovalRequired(value: unknown): value is ApprovalRequiredOutput {
  if (!isObject(value)) return false;
  if (value.code !== AgentGateCode.ApprovalRequired) return false;
  if (!AGENT_GATE_CODES.has(value.code as AgentGateCode)) return false;
  return readStringField(value, "approval_id") !== null
    && readStringField(value, "expires_at") !== null
    && readStringField(value, "message") !== null;
}

export function isMocksExhausted(value: unknown): value is MocksExhaustedOutput {
  if (!isObject(value)) return false;
  if (value.code !== AgentGateCode.MocksExhausted) return false;
  if (!AGENT_GATE_CODES.has(value.code as AgentGateCode)) return false;
  return readStringField(value, "message") !== null;
}

export function isMocksEngineError(value: unknown): value is MocksEngineErrorOutput {
  if (!isObject(value)) return false;
  if (value.code !== AgentGateCode.MocksEngineError) return false;
  if (!AGENT_GATE_CODES.has(value.code as AgentGateCode)) return false;
  return readStringField(value, "message") !== null;
}

export function isToolNotMocked(value: unknown): value is ToolNotMockedOutput {
  if (!isObject(value)) return false;
  if (value.code !== AgentGateCode.ToolNotMocked) return false;
  if (!AGENT_GATE_CODES.has(value.code as AgentGateCode)) return false;
  return readStringField(value, "tool_name") !== null
    && readStringField(value, "message") !== null;
}

/**
 * Exhaustive-match witness. A `switch` over `AgentGateCode` should
 * pass `value` here in the default branch — TS fails to compile if
 * `value` isn't narrowed to `never`, i.e. if a code variant escaped
 * the switch. The runtime body throws so a hostile cast at runtime
 * doesn't silently swallow an unknown code.
 */
export function assertExhaustiveAgentGate(value: never): never {
  throw new Error(
    `agent-gate: unexpected output variant ${JSON.stringify(value)}`,
  );
}
